src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
+        }
+
+        @Override
+        public DuplicationMetrics decode( InputStream inStream, Context context ) throws CoderException, IOException {
+            DuplicationMetrics metrics = new DuplicationMetrics();
+            metrics.LIBRARY = SerializableCoder.of(String.class).decode(inStream, context);
+            metrics.UNMAPPED_READS = SerializableCoder.of(Long.class).decode(inStream, context);
+            metrics.UNPAIRED_READS_EXAMINED = SerializableCoder.of(Long.class).decode(inStream, context);
+            metrics.READ_PAIRS_EXAMINED = SerializableCoder.of(Long.class).decode(inStream, context);
+            metrics.UNPAIRED_READ_DUPLICATES = SerializableCoder.of(Long.class).decode(inStream, context);
+            metrics.READ_PAIR_DUPLICATES = SerializableCoder.of(Long.class).decode(inStream, context);
+            metrics.READ_PAIR_OPTICAL_DUPLICATES = SerializableCoder.of(Long.class).decode(inStream, context);
+            metrics.PERCENT_DUPLICATION = SerializableCoder.of(Double.class).decode(inStream, context);
+            metrics.ESTIMATED_LIBRARY_SIZE = SerializableCoder.of(Long.class).decode(inStream, context);
+            return metrics;
+        }
+    }
+
+    private static void writeMetricsToFile(Pipeline pipeline, final PCollection<KV<String,DuplicationMetrics>> metrics,
+                                           final SAMFileHeader header, final File dest) {
+        PCollectionView<Map<String,Iterable<DuplicationMetrics>>> mapView = metrics
+            .apply(View.asMap());
+
+        PCollection<File> dummy = pipeline.apply(Create.<File>of(dest).setName("output metrics file"));
+
+        dummy.apply(ParDo.named("save metrics file")
+                    .withSideInputs(mapView)
+                    .of(new WriteToMetricsFile(header, mapView))
+        );
+
+    }
+
+    private static class WriteToMetricsFile extends DoFn<File,Void> implements Serializable {
+        private static final long serialVersionUID = 1L;
+        private final SAMFileHeader header;
+        private final PCollectionView<Map<String,Iterable<DuplicationMetrics>>> mapView;
+
+        public WriteToMetricsFile(SAMFileHeader header, PCollectionView<Map<String,Iterable<DuplicationMetrics>>> mapView) {
+            this.header = header;
+            this.mapView = mapView;
+        }
+
+        @Override
+        public void processElement(ProcessContext c) throws Exception {
+            File dest = c.element();
+            Map<String,Iterable<DuplicationMetrics>> metrics = c.sideInput(mapView);
+            final MetricsFile<DuplicationMetrics,Double> file = new MetricsFile<>();
+            for (final Map.Entry<String,Iterable<DuplicationMetrics>> entry : metrics.entrySet()) {
+                for (final DuplicationMetrics m : entry.getValue()) {
+                    file.addMetric(m);
+                }
+            }
+            // Should we add a histogram?
```

Comment:
the way the metrics files are _currently_ setup there can only be one histogram (though it can have multiple values) and the headers are fixed in advance. This means that when multiple libraries are present we cannot display the detailed metrics (which are per library)...It would be nice to have a solution that can support multiple histograms in the result. @ktibbett perhaps you would like to chime in as well?